pkg/queue/instance.go

```diff
@@ -66,10 +66,21 @@ func (q *queueImpl) Run(stop <-chan struct{}) {
 		q.cond.L.Unlock()
 	}()
 
+	events := make(chan struct{})
 	for {
 		q.cond.L.Lock()
 		for !q.closing && len(q.tasks) == 0 {
-			q.cond.Wait()
+			go func() {
+				q.cond.Wait()
+				events <- struct{}{}
+			}()
+
+			select {
+			case <-stop:
+				return
+			case <-events:
+				// do task
+			}
```

Comment:
How about 
```
		for !q.closing && len(q.tasks) == 0 {
			select {
			case <-stop:
				break
			default:
				q.cond.Wait()
			}
		}
```

Reply:
never mind.

Reply:
if block `q.cond.Wait()`, the problem not resolve

Reply:
```
	go func() {
		<-stop
		q.cond.L.Lock()
                q.cond.Signal()
		q.closing = true
		q.cond.L.Unlock()
	}()

	for {
		q.cond.L.Lock()
		for !q.closing && len(q.tasks) == 0 {
			q.cond.Wait()
		}
```

Reply:
it work well, and cover 100%